// ng-password-validator/esm5/lib/ng-password-validator.module.js
import { __decorate } from "tslib";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { DataService } from "./data.service";
import { NgPasswordValidatorComponent } from "./ng-password-validator.component";
import { NgPasswordValidatorDirective } from "./ng-password-validator.directive";
import { NgPasswordValidatorService } from "./ng-password-validator.service";
var NgPasswordValidatorModule = /** @class */ (function () {
    function NgPasswordValidatorModule() {
    }
    NgPasswordValidatorModule_1 = NgPasswordValidatorModule;
    /**
     * Password validator module
     *
     * @static
     * @param {NgPasswordValidatorOptions} initOptions
     * @returns {ModuleWithProviders<NgPasswordValidatorModule>}
     * @memberof NgPasswordValidatorModule
     */
    NgPasswordValidatorModule.forRoot = function (initOptions) {
        return {
            ngModule: NgPasswordValidatorModule_1,
            providers: [
                {
                    provide: NgPasswordValidatorService,
                    useValue: initOptions
                }
            ]
        };
    };
    var NgPasswordValidatorModule_1;
    NgPasswordValidatorModule = NgPasswordValidatorModule_1 = __decorate([
        NgModule({
            declarations: [
                NgPasswordValidatorDirective,
                NgPasswordValidatorComponent,
            ],
            imports: [
                CommonModule
            ],
            providers: [DataService],
            exports: [
                NgPasswordValidatorDirective
            ],
            entryComponents: [
                NgPasswordValidatorComponent
            ]
        })
    ], NgPasswordValidatorModule);
    return NgPasswordValidatorModule;
}());
export { NgPasswordValidatorModule };
